### PR TITLE
Supporting Timestamp with microseconds precision

### PR DESCRIPTION
**Pull request by alextekartik**
So I no longer force timestamp on the js side and I fixed JsDate parsing so that existing code and tests work without changes. User will see a warning unless he calls explicitely settings with `timestampsInSnapshots: true`. This will enforce the developer to migrate to it. When using it, dates/timestamps are properly read as timestamp with microseconds precision

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,6 @@ doc/api/
 # Files with private data. Used for development and testing
 lib/src/assets/config.json
 lib/src/assets/service_account.json
+
+# IDE
+.idea/
```
Comment (kevmoo) [inline .gitignore]: Please, no. These should be ignored globally.

File: test/firestore_test.dart [hunk 2]
```diff
@@ -55,8 +55,7 @@ void main() {
         databaseURL: databaseUrl,
         projectId: projectId,
         storageBucket: storageBucket);
-
-    firestore = fb.firestore();
+    firestore = app.firestore();
```
Comment (kevmoo) [inline test/firestore_test.dart]: Why this change?

File: lib/src/firestore.dart [hunk 3]
```diff
@@ -33,8 +33,7 @@ class Firestore extends JsObjectWrapper<firestore_interop.FirestoreJsImpl> {
       return null;
     }
 
-    return _expando[jsObject] ??= Firestore._fromJsObject(jsObject
-      ..settings(firestore_interop.Settings(timestampsInSnapshots: true)));
```
Comment (kevmoo) [inline lib/src/firestore.dart]: Uh...don't we want this to stay here?

File: .gitignore (hunk 1 above)
Comment (alextekartik) [reply .gitignore]: Oups sorry. Will revert

File: test/firestore_test.dart (hunk 2 above)
Comment (alextekartik) [reply test/firestore_test.dart]: Actually I made this change first because in my testing I was initializing another app first and fb.firestore() seems to pick the first one. So I had to explicitly make it point to the correct app and I thought it was more correct to keep it this way. I initialize another app in the same file so I found it clearer to specify the app from which firestore is picked. Do you want me to revert this change?

File: lib/src/firestore.dart (hunk 3 above)
Comment (alextekartik) [reply lib/src/firestore.dart]: That’s actually a breaking change now to call `timestampsInSnapshots: true`. The default as specified in the documentation is to have Date in snapshot. There was a bug in dartifyDate that prevented this to work properly that I fixed.

Initially I wanted to add another settings to allow using DateTime or Timestamp at the app level, but I could not manage to have dartify work differently in the 2 usages depending on the firestore app used and actually the existing setting works perfectly (we get js Date or js Timestamp so it easy to do the proper conversion in dartify).

It was used the wrong way: we were forcing to use Timestamp and converting every timestamp to Date loosing the microsecond precision, it seems it was just here to hide a warning that is actually relevant. If we decide to make  `timestapsInSnapshots: true` the default, that will break existing apps and I don’t want to be the one changing this (although I’m all in favor). The warning in the console is a good step to have developers migrating to Timestamp which might be the only option in the future.

As shown here (https://github.com/tekartik/firebase-dart/blob/ebc04f57317cbe380eff14f4c6d678f1b37c1724/test/firestore_test.dart#L976-L1015), when you use `timestampsInSnapshots: true` you only get timestamp. However you can write using either DateTime or Timestamp

File: .gitignore (hunk 1 above)
Comment (alextekartik) [reply .gitignore]: reverted

**Comment by alextekartik** [comment]
Should I cancel this pull request? I agree Timestamp is not a priority on the Web however we loose the microsecond precision since every timestamp is converted to Date.

**Comment by kevmoo** [comment]
I noticed that latest API update prints a bunch of errors here.

Strictly speaking, we should bump this to a major release if we're going to remove this. 

Want to rebase on latest and I'll look again?

**Comment by kevmoo** [comment]
Replaced by #208 ?
